tests/operations.go

```diff
@@ -184,3 +202,92 @@ func isNotFound(err error) bool {
 		return k8serrors.IsNotFound(err)
 	}
 }
+
+// GetHealthyCSPCount ...
+func (ops *Operations) GetHealthyCSPCount(spcName string) int {
+	cspAPIList, err := ops.CSPClient.List(metav1.ListOptions{})
+	Expect(err).To(BeNil())
+	count := csp.
+		ListBuilderForAPIObject(cspAPIList).
+		List().
+		Filter(csp.HasLabel(string(apis.StoragePoolClaimCPK), spcName), csp.IsStatus("Healthy")).Len()
+	return count
+}
+
+// DeleteCSP ...
+func (ops *Operations) DeleteCSP(spcName string, deleteCount int) {
+	cspAPIList, err := ops.CSPClient.List(metav1.ListOptions{})
+	Expect(err).To(BeNil())
+	cspList := csp.
+		ListBuilderForAPIObject(cspAPIList).
+		List().
+		Filter(csp.HasLabel(string(apis.StoragePoolClaimCPK), spcName), csp.IsStatus("Healthy"))
+	cspCount := cspList.Len()
+	Expect(deleteCount).Should(BeNumerically("<=", cspCount))
+
+	for i := 0; i < deleteCount; i++ {
+		_, err := ops.CSPClient.Delete(cspList.ObjectList.Items[i].Name, &metav1.DeleteOptions{})
+		Expect(err).To(BeNil())
+
+	}
+}
+
+// IsHealthyCspCount ...
+func (ops *Operations) IsHealthyCspCount(spcName string, expectedCspCount int) int {
+	var cspCount int
+	retries := maxRetry
+	for i := 0; i < retries; i++ {
+		cspCount = ops.GetHealthyCSPCount(spcName)
+		if cspCount == expectedCspCount {
+			return expectedCspCount
+		}
+		if retries == 0 {
+			break
+		}
+		retries--
+		time.Sleep(5 * time.Second)
+	}
+	return cspCount
+}
+
+// ExecPod executes arbitrary command inside the pod
+func (ops *Operations) ExecPod(podName, namespace, containerName string, command ...string) ([]byte, error) {
```

Comment:
These methods of operations is giving rise to an anti pattern.
I felt, the test related business logic is getting split across multiple places.
In other words, test logic is no longer cohesive.
We are trying to make the common logic modular & at the same time losing the cohesion.

Reply:
### Issues without a proper ops approach -- some observations
- Rise of utils
- For example, below code snippets are used during integration tests
  - This core testing logic is placed in different file/struct
  - The original test logic is somewhere else
  - This leads to splitting of test related business logic across multiple files
  - In other words, test logic is not cohesive
  - Scattering of core test logic can be verified from
    - Methods names that do not reflect its true behaviour
    - Explosion of arguments for a single method
    - Once can also find test logic is tightly coupled with `Expect` statements
- In other words, this test logic is not usable in other scenarios
  - e.g. monitoring, self-heal, etc.

Reply:
This is done, so that other tests can also reuse the same to avoid repetition at various places. What is the benifit if we keep repeting the the same func/methods for different test components, like we are doing for pod/service/deployment. I see the same code has been repeated in for every kubernetes resources.

Reply:
Refactoring of bdd framework will be done in other PR.